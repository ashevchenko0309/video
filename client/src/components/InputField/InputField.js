import React, { useState, useCallback } from 'react';
import { throttle } from 'lodash';
import classNames from 'classnames';

import FormFieldErrorHint from '../FormErrorHints/FormFieldErrorHint';

function InputField({
  type,
  name,
  defaultValue,
  placeholder,
  labelText,
  isRequired,
  options,
  hasError,
  delay,
  errorMessage,
  onChange,
  onBlur,
}) {

  const [inputValue, setInputValue] = useState(defaultValue);

  const throttledInputChange = useCallback(throttle(v => onChange(v), delay), []);

  const _options = options || { minLength: isRequired ? 1 : 0, maxLength: 120 };

  const formFieldClass = classNames({
    "form_field": true,
    "has-error": hasError
  })

  const _onChange = (e) => {
    const value = e.currentTarget.value;
    setInputValue(value);
    throttledInputChange(value);
  }

  const _onBlur = (e) => {
    const value = e.currentTarget.value;
    onBlur(value, name);
  }

  return (
    <div className={formFieldClass}>
      <label htmlFor={name}>
        <span className="input_field--label">{labelText}{isRequired && <span className="is-required"> *</span>}</span>
        <span className="input_field--hint">{inputValue.length} / {_options.maxLength}</span>
        <input
          type={type}
          className="form_field--input"
          value={inputValue}
          name={name}
          placeholder={placeholder}
          required={isRequired}
          maxLength={_options.maxLength}
          onChange={_onChange}
          onBlur={_onBlur}
        />
      </label>
      <FormFieldErrorHint hasError={hasError} errorMessage={errorMessage} />
    </div>
  )

}

InputField.defaultProps = {
  type: "text",
  name: '',
  defaultValue: '',
  placeholder: '',
  labelText: '',
  isRequired: false,
  options: null,
  hasError: false,
  delay: 150,
  errorMessage: '',
  onChange: () => { },
  onBlur: () => { }
}

export default InputField;