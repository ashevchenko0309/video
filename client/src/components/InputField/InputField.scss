.form_field--input{
  display: block;
  width: 100%;
}

.input_field--hint{
  float: right;
}

.input_field--error{
  color: rgb(194, 17, 17);
  &.inactive{
    display: none !important;
  }
  &.active{
    display: initial;
  }
}