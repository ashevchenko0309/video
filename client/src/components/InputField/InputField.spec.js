import React from "react";
import {create, act} from 'react-test-renderer';
import InputField from "./InputField";

const DEFAULT_PROPS = { name: "Test", placeholder: "Test placeholder", labelText: "Test label" }

const setUp = (props) => mount(<InputField {...props} />);

describe("InputField component", () => {

  let component;

  beforeEach(() => {
    component = setUp({ ...DEFAULT_PROPS });
  });

  it("should render default input (snapshot)", () => {
    expect(component.getElement()).toMatchSnapshot();
  });

  it("should render input with default value (snapshot)", () => {
    const component = setUp({ ...DEFAULT_PROPS, defaultValue: "default value" })
    expect(component.getElement()).toMatchSnapshot();
  });

  it("should render input with error message (snapshot)", () => {
    const component = setUp({ ...DEFAULT_PROPS, hasError: true, errorMessage: "Some error" })
    expect(component.getElement()).toMatchSnapshot();
  });

  it("should render textarea with all props (snapshot)", () => {
    const component = setUp({ ...DEFAULT_PROPS, defaultValue: "default value", hasError: true, errorMessage: "Some error" })
    expect(component.getElement()).toMatchSnapshot();
  });

  it("should have default length hint", () => {
    expect(component.find('.input_field--hint').text()).toBe("0 / 120");
  });

  it("should have required hint", () => {
    const inputField = setUp({ ...DEFAULT_PROPS, isRequired: true });
    expect(inputField.find('.input_field--label').text()).toBe("Test label *");
  });

  it("should have default inactive class error hint", () => {
    const inputField = setUp({ ...DEFAULT_PROPS, isRequired: true });
    expect(inputField.find('.input_field--error').hasClass('inactive')).toBe(true);
  });

  it("should have active class error hint", () => {
    const inputField = setUp({ ...DEFAULT_PROPS, isRequired: true, hasError: true, errorMessage: 'Test error' });
    expect(inputField.find('.form_field').hasClass('has-error')).toBe(true);
    expect(inputField.find('.input_field--error').hasClass('active')).toBe(true);
  });

  it("should have error hint", () => {
    const inputField = setUp({ ...DEFAULT_PROPS, isRequired: true, hasError: true, errorMessage: 'Test error' });
    expect(inputField.find('.input_field--error').text()).toBe("Test error");
  });

  it("should call onChange method", () => {
    const mockCallback = jest.fn();
    const inputField = setUp({ ...DEFAULT_PROPS, onChange: mockCallback });

    expect(mockCallback.mock.calls.length).toBe(0);
    inputField.find('.form_field--input').simulate('change', { currentTarget: { value: '123456789' } });
    expect(mockCallback.mock.calls.length).toBe(1);
  });

  it("shold call setInputValue method", () => {
    const setInputValue = jest.fn();
    const inputField = setUp({ ...DEFAULT_PROPS, onChange: setInputValue });
    const handleType = jest.spyOn(React, "useState");
    handleType.mockImplementation(inputValue => [inputValue, setInputValue]);

    expect(setInputValue.mock.calls.length).toBe(0);

    inputField.find('.form_field--input').simulate('change', { currentTarget: { value: '123456789' } });
    expect(setInputValue).toBeTruthy();
    expect(setInputValue.mock.calls.length).toBe(1);
  })

  it("should call onBlur method", () => {
    const onBlur = jest.fn();
    const inputField = setUp({ ...DEFAULT_PROPS, onBlur });

    expect(onBlur.mock.calls.length).toBe(0);

    inputField.find('.form_field--input').simulate('blur', { currentTarget: { value: '123456789' } });
    expect(onBlur.mock.calls.length).toBe(1);
  });

  it("should use default onChange prop", () => {
    const result = InputField.defaultProps.onChange();
    expect(result).toBe(undefined);
  });

  it("should use default onBlur prop", () => {
    const result = InputField.defaultProps.onBlur();
    expect(result).toBe(undefined);
  });

});
