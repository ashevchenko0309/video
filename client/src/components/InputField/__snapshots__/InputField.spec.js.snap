// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputField component should render default input (snapshot) 1`] = `
<InputField
  defaultValue=""
  delay={150}
  errorMessage=""
  hasError={false}
  isRequired={false}
  labelText="Test label"
  name="Test"
  onBlur={[Function]}
  onChange={[Function]}
  options={null}
  placeholder="Test placeholder"
  type="text"
/>
`;

exports[`InputField component should render input with default value (snapshot) 1`] = `
<InputField
  defaultValue="default value"
  delay={150}
  errorMessage=""
  hasError={false}
  isRequired={false}
  labelText="Test label"
  name="Test"
  onBlur={[Function]}
  onChange={[Function]}
  options={null}
  placeholder="Test placeholder"
  type="text"
/>
`;

exports[`InputField component should render input with error message (snapshot) 1`] = `
<InputField
  defaultValue=""
  delay={150}
  errorMessage="Some error"
  hasError={true}
  isRequired={false}
  labelText="Test label"
  name="Test"
  onBlur={[Function]}
  onChange={[Function]}
  options={null}
  placeholder="Test placeholder"
  type="text"
/>
`;

exports[`InputField component should render textarea with all props (snapshot) 1`] = `
<InputField
  defaultValue="default value"
  delay={150}
  errorMessage="Some error"
  hasError={true}
  isRequired={false}
  labelText="Test label"
  name="Test"
  onBlur={[Function]}
  onChange={[Function]}
  options={null}
  placeholder="Test placeholder"
  type="text"
/>
`;
